gossipsub.go

```diff
@@ -239,6 +264,90 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 			gs.tracer.Prune(p, topic)
 			delete(peers, p)
 			gs.untagPeer(p, topic)
+			gs.addBackoff(p, topic)
```

Comment:
👍 great!

However, we're not currently doing anything about a peer continuously chasing away other peers by sending a GRAFT. If during a GRAFT we could check if we're already over the limit and send pack a PRUNE that would largely resolve this case.

Reply:
We don't want to do that because we will reject all new peers and they won't be able to form a connected mesh.
That's why it accepts the peer and resolves (with randomization) during the heartbeat.

Reply:
Just to be clear, sending a PRUNE if we are over the limit is the wrong thing to do, because then the mesh can become full and fail to accept new peers.
(I considered it during when designing the algorithm)

Reply:
I get that it used to be that way, but is that still true now that we have peer exchange?

If I send back a PRUNE only when I have >Dhi peers and I prune (Dhi-D) peers and tell them about each other won't they always be able to join the mesh?

Reply:
They probably will be able to do that, but I'd rather reshuffle the mesh in its entirety instead of rejecting new GRAFTs.

Reply:
Think about a fully connected mesh where all peers are at `D_hi` -- (unlikely as it is) it won't accept new peers then, while GRAFTing and reshuffling will resolve the situation.

Reply:
Ah, so IIUC this is really a timing problem right? Because if B is connected to C and then A tries to connect to B it's possible that when B PRUNEs A and C that A will send a GRAFT to C **before** C receives the PRUNE from B.

My concern with the current approach is that if A wants to GRAFT to B it can always keep spamming B and eventually it will get through, even though B has given A plenty of other peers to connect to. This is annoying since if a peer is "better" in some way (e.g. they are the primary publisher) then nodes might be selfish, however it's certainly not a deal-breaker and wasn't even resolvable until the Peer Exchange changes.

Since fixing this would likely require some protocol thought and is less important than the rest of the episub work, seems reasonable to push this further down the road. Would you like me to create an issue?

Reply:
It could be an attack -- I don't think it can happen naturally.

Reply:
Sure, we can create an issue to discuss further instead of it being lost in oblivion after the pr merges.

Reply:
There's a recent win here! The decisions of which peers to retain and which to eject is now performed by evaluating the peer's score! 🎉